src/handle_wrap.cc

```diff
@@ -93,14 +93,63 @@ void HandleWrap::OnClose(uv_handle_t* handle) {
   Context::Scope context_scope(env->context());
   Local<Object> object = wrap->object();
 
-  if (wrap->flags_ & kCloseCallback) {
+  if (wrap->flags_ & kCloseCallback)
     wrap->MakeCallback(env->onclose_string(), 0, nullptr);
+
+  if (wrap->flags_ & kEnvironmentCleanup) {
+    CHECK_EQ(wrap->hc_, nullptr);
+    env->FinishHandleCleanup(handle);
+  } else if (wrap->hc_ != nullptr) {
+    // Prevent double clean ups.
+    env->DeregisterHandleCleanup(wrap->hc_);
+    wrap->hc_ = nullptr;
   }
 
-  object->SetAlignedPointerInInternalField(0, nullptr);
+  node::ClearWrap(object);
   wrap->persistent().Reset();
+  handle->data = nullptr;
   delete wrap;
 }
 
 
+void HandleWrap::HandleCleanupCallback(Environment* env,
+                                       uv_handle_t* handle,
+                                       void* arg) {
+  if (handle->data == nullptr)
+    return;
+
+  HandleWrap* wrap = static_cast<HandleWrap*>(handle->data);
+  // At this point Environment has freed the HandleCleanup* and this is the
+  // first opportunity to nullify it.
+  // TODO(petkaantonov) it would be more symmetric with DeregisterHandleCleanup
+  // if the HandleCleanup* was deleted in FinishHandleCleanup but it requires
+  // changing all other users of RegisterHandleCleanup
+  wrap->hc_ = nullptr;
+  if (handle != nullptr && !uv_is_closing(handle)) {
+    if (arg != nullptr) {
+      CallbackContainer* container = static_cast<CallbackContainer*>(arg);
```

Comment:
This implements `HandleCleanup` that has the signature `Environment*, uv_handle_t*, void*`